query/src/main/java/org/infinispan/query/impl/IndexStartupRunner.java

```diff
@@ -55,6 +58,46 @@ private static IndexStartupMode computeFinalMode(Configuration configuration) {
             }
          }
       }
+
+      if (DataKind.SHARED_MEMORY_MAX_COUNT.equals(dataKind) && !indexesAreVolatile &&
+            IndexStartupMode.AUTO.equals(startupMode)) {
+         // in this case a node that is crashed could have not aligned indexes.
+         // we force the reindex when it recovers to be sure that is aligned.
+         return IndexStartupMode.REINDEX;
+      }
+
       return (IndexStartupMode.AUTO.equals(startupMode)) ? IndexStartupMode.NONE : startupMode;
    }
+
+   private enum DataKind {
+      VOLATILE, PERSISTENT, SHARED_MEMORY_MAX_COUNT
+   }
+
+   private static DataKind computeDataKind(Configuration configuration) {
+      List<StoreConfiguration> cacheStores = configuration.persistence().stores();
+      if (cacheStores.isEmpty()) {
+         return DataKind.VOLATILE;
+      }
+
+      boolean nonSharedCacheStore = false;
+      boolean sharedCacheStore = false;
+      for (StoreConfiguration cacheStore : cacheStores) {
+         if (cacheStore.purgeOnStartup()) {
+            continue;
+         }
+         if (cacheStore.shared()) {
+            sharedCacheStore = true;
+         } else {
+            nonSharedCacheStore = true;
+         }
+      }
+
+      if (nonSharedCacheStore) {
+         return DataKind.PERSISTENT;
+      }
+      if (sharedCacheStore && configuration.memory().maxCount() > -1L) {
```

Comment:
Is it only maxCount?
If maxSize has been set it is still eviction=enabled and the index need reindex as well?
So it is not SHARED_MEMORY_MAX_COUNT but SHARED_WITH_EVICTION ?

Reply:
@wfink is right. I'll make it simpler: if a shared store is configured, it must re-index the data to be on the safer side.

Reply:
You're right, adding it! Thanks

Reply:
@pruivo if `isEvictionEnabled` is false is it possible to have this issue?

Reply:
could you reproduce those other cases with a test? I'm asking since I want to limit the redindex at startup as much as we can.

Reply:
If a user calls `evict` explicitly using indexes and a shared store, and a node crashes and recover, he deserves to have not aligned indexes .) I don't want to add more reindex on normal case to support this very peculiar and unsafe case.

Reply:
Added both combinations: https://github.com/pruivo/infinispan/blob/a513e727b5284c22a9fb6f758e0c39c80a969591/query/src/test/java/org/infinispan/query/distributed/IndexWithSharedStoreTest.java

`cache.evict()` is not unsafe and it is used by users who want to customize the eviction strategy.

Reply:
At most we can add a note on the documetation saying that if the user does explicit eviction (only possible with embedded caches I think) if indexes are used, if shared storages are used if a node crash, the index could be not aligned. But I would avoid to support it with the auto, adding non required reindex at startup for more cases, where those are not required.

Reply:
https://infinispan.org/docs/stable/titles/configuring/configuring.html#eviction-manual_configuring-memory-usage

Reply:
Is this supported by remote API?